### PR TITLE
[SCB-1047]microservice.yaml  service_description.version support form…

### PR DESCRIPTION
**Pull request by heyile**
…at xxx.xx.xxx.xxx

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: core/src/test/java/org/apache/servicecomb/core/TestConfig.java [hunk 1]
```diff
@@ -53,7 +53,7 @@ public void testConstants() {
     Assert.assertEquals("rest", Const.RESTFUL);
     Assert.assertEquals("", Const.ANY_TRANSPORT);
     Assert.assertEquals("latest", Const.VERSION_RULE_LATEST);
-    Assert.assertEquals("0.0.0+", Const.DEFAULT_VERSION_RULE);
+    Assert.assertEquals("0.0.0.0+", Const.DEFAULT_VERSION_RULE);
```
Comment (liubao68) [inline core/src/test/java/org/apache/servicecomb/core/TestConfig.java]: Implement this feature, we must make sure the old version rule works. Can you add new test cases for 4 bit version with change any old test case?

Comment (heyile) [reply core/src/test/java/org/apache/servicecomb/core/TestConfig.java]: As the const DefinitionConst#VERSION_RULE_ALL  is changed from "0.0.0+" to "0.0.0.0+",
I change the value of Const.DEFAULT_VERSION_RULE. you means that I shouldn't change the const 
DefinitionConst#VERSION_RULE_ALL ?

Comment (liubao68) [reply core/src/test/java/org/apache/servicecomb/core/TestConfig.java]: I think we do not need to change the VERSION_RULE_ALL constant..

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java [hunk 2]
```diff
@@ -23,35 +23,44 @@
...
 
   private final short major;
 
   private final short minor;
 
   private final short patch;
 
+  private final short build;
+
+  private final boolean versionFour;
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: why need a flag versionFour?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java [hunk 3]
```diff
@@ -53,8 +54,11 @@ private Microservice createMicroserviceFromDefinition(Configuration configuratio
     microservice.setServiceName(configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_NAME_KEY,
         DEFAULT_MICROSERVICE_NAME));
     microservice.setAppId(configuration.getString(CONFIG_APPLICATION_ID_KEY, DEFAULT_APPLICATION_ID));
-    microservice.setVersion(configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_VERSION_KEY,
-        DEFAULT_MICROSERVICE_VERSION));
+    String version = configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_VERSION_KEY,
+        DEFAULT_MICROSERVICE_VERSION);
+    //check version format
+    VersionRuleUtils.checkVersionFormat(version);
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: versionRule to check version?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java [hunk 4]
```diff
@@ -54,7 +54,17 @@ public static VersionRule create(String strVersionRule) {
         return versionRule;
       }
     }
+    throw new IllegalStateException("config service_description.version is invalid ");
+  }
 
-    throw new IllegalStateException("never run to here");
+  public static void checkVersionFormat(String strVersionRule) {
+    strVersionRule = strVersionRule.trim();
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java]: named check version, but checking versionRule?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java [hunk 5]
```diff
@@ -21,37 +21,41 @@
...
     String[] versions = version.split("\\.", -1);
-    if (versions.length > 3) {
+    if (versions.length > 4) {
       throw new IllegalStateException(String.format("Invalid version \"%s\".", version));
     }
 
-    if (versions.length < 3) {
-      versions = (String[]) ArrayUtils.addAll(versions, ZERO);
-    }
+    versions = (String[]) ArrayUtils.addAll(versions, ZERO);
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: less then 4, need to concat array

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java [hunk 6]
```diff
@@ -74,34 +78,52 @@ private short parseNumber(String name, String allVersion, String version) {
...
     this.minor = minor;
     this.patch = patch;
+    this.build = build;
     this.version = combineStringVersion();
     this.numberVersion = combineVersion();
   }
 
   private String combineStringVersion() {
-    return major + "." + minor + "." + patch;
+    StringBuilder stringBuilder = new StringBuilder();
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: performance same to add directly

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java [hunk 7]
```diff
@@ -74,34 +78,52 @@ private short parseNumber(String name, String allVersion, String version) {
...
+    return stringBuilder.toString();
   }
 
+  // 1.0.0 equals 1.0.0.0
   private long combineVersion() {
-    return (long) major << 32 | (long) minor << 16 | (long) patch;
+    return (long) major << 48 | (long) minor << 32 | (long) patch << 16 | (long) build;
   }
 
+  @VisibleForTesting
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: why VisibleForTesting?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java [hunk 8]
```diff
@@ -53,8 +54,13 @@ private Microservice createMicroserviceFromDefinition(Configuration configuratio
     microservice.setServiceName(configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_NAME_KEY,
         DEFAULT_MICROSERVICE_NAME));
     microservice.setAppId(configuration.getString(CONFIG_APPLICATION_ID_KEY, DEFAULT_APPLICATION_ID));
-    microservice.setVersion(configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_VERSION_KEY,
-        DEFAULT_MICROSERVICE_VERSION));
+    String version = configuration.getString(CONFIG_QUALIFIED_MICROSERVICE_VERSION_KEY,
+        DEFAULT_MICROSERVICE_VERSION);
+    //check version format
+    if (!Version.checkVersion(version)) {
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: why not just new a Version instance

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java [hunk 9]
```diff
@@ -54,7 +54,6 @@ public static VersionRule create(String strVersionRule) {
         return versionRule;
       }
     }
-
-    throw new IllegalStateException("never run to here");
+    throw new IllegalStateException("config service_description.version is invalid ");
```
Comment (wujimin) [inline service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java]: why versionRule relate to "service_description.version"?

File: service-registry/src/test/java/org/apache/servicecomb/serviceregistry/registry/TestRemoteServiceRegistry.java [hunk 10]
```diff
@@ -141,6 +142,13 @@ public void onPullMicroserviceVersionsInstancesEvent(@Injectable ServiceRegistry
       }
     }.getMockInstance();
 
+    new MockUp<Version>() {
+      @Mock
+      boolean checkVersion(String version){
```
Comment (wujimin) [inline service-registry/src/test/java/org/apache/servicecomb/serviceregistry/registry/TestRemoteServiceRegistry.java]: why need to mock a valid version?

Comment (heyile) [reply service-registry/src/test/java/org/apache/servicecomb/serviceregistry/registry/TestRemoteServiceRegistry.java]: as now when version is invalid, we can not registe provider

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java (hunk 5 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: ok

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java (hunk 7 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: it's only used in Junit

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java (hunk 9 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/VersionRuleUtils.java]: I mistake version and versionRule at first. And forget to delete it

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java (hunk 7 above)
Comment (wujimin) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: it's a public class, and have normal getter/setter?
why only can be used in junit?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java (hunk 8 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: Maybe just check format is better ? the new Version instance will never used

Comment (wujimin) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: but constrcut a version will got the detail of the wrong format message

Comment (wujimin) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: do you want to express version format check logic by two different way?

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java (hunk 7 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/version/Version.java]: ok

File: service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java (hunk 8 above)
Comment (heyile) [reply service-registry/src/main/java/org/apache/servicecomb/serviceregistry/api/registry/MicroserviceFactory.java]: ok, I see.

**Pull request merged**